core/primitives/src/utils.rs

```diff
@@ -25,50 +25,108 @@ pub mod min_heap;
+    payload: T,
 }
 
 impl<T> MaybeValidated<T> {
-    pub fn validate_with<E, F: FnOnce(&T) -> Result<bool, E>>(&self, f: F) -> Result<bool, E> {
-        match &self {
-            Self::Validated(_) => Ok(true),
-            Self::NotValidated(t) => f(t),
-        }
+    /// Creates new MaybeValidated object marking payload as validated.  No
+    /// verification is performed; it’s caller’s responsibility to make sure the
+    /// payload was indeed validated.
+    pub fn from_validated(payload: T) -> Self {
+        Self { validated: std::cell::Cell::new(true), payload }
     }
 
-    pub fn map<U, F: FnOnce(T) -> U>(self, f: F) -> MaybeValidated<U> {
-        match self {
-            Self::Validated(t) => MaybeValidated::Validated(f(t)),
-            Self::NotValidated(t) => MaybeValidated::NotValidated(f(t)),
+    /// Validates payload with given `validator` function and returns result of
+    /// the validation.  If payload has already been validated returns
+    /// `Ok(true)`.  Note that this method changes the internal validated flag
+    /// so it’s probably incorrect to call it with different `validator`
+    /// functions.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use near_primitives::utils::MaybeValidated;
+    ///
+    /// let value = MaybeValidated::from(42);
+    /// assert_eq!(Ok(true), value.validate_with::<(), _>(|v| Ok(*v == 42)));
+    /// assert_eq!(Ok(true), value.validate_with::<(), _>(|_| panic!()));
+    /// ```
+    pub fn validate_with<E, F: FnOnce(&T) -> Result<bool, E>>(
+        &self,
+        validator: F,
+    ) -> Result<bool, E> {
+        if self.validated.get() {
+            Ok(true)
+        } else {
+            let res = validator(&self.payload);
+            self.validated.set(*res.as_ref().unwrap_or(&false));
+            res
         }
     }
 
+    /// Applies function to the payload (whether it’s been validated or not) and
+    /// returns new object with result of the function as payload.  Validated
+    /// state is not changed.
+    pub fn map<U, F: FnOnce(T) -> U>(self, validator: F) -> MaybeValidated<U> {
+        MaybeValidated { validated: self.validated, payload: validator(self.payload) }
+    }
+
+    /// Returns a new object storing reference to this object’s payload.  Note
+    /// that the two objects do not share the validated state so calling
+    /// `validate_with` on one of them does not affect the other.
     pub fn as_ref(&self) -> MaybeValidated<&T> {
-        match &self {
-            Self::Validated(ref t) => MaybeValidated::Validated(t),
-            Self::NotValidated(ref t) => MaybeValidated::NotValidated(t),
-        }
+        MaybeValidated { validated: self.validated.clone(), payload: &self.payload }
     }
 
-    pub fn extract(self) -> T {
-        match self {
-            Self::Validated(t) => t,
-            Self::NotValidated(t) => t,
-        }
+    /// Extracts the payload whether or not it’s been validated.
+    pub fn into_inner(self) -> T {
```

Comment:
Does it make sense to extract only when it's validated? I think the existing use case for PartialEncodedChunk validation satisfies that we only want to extra the inner struct only if it's validated. Obviously you know better for the use cases in `process_block`, but I think it makes sense that we only want to process a block if the block has already been validated.

Reply:
I don’t think so. The use in `ShardsManager::process_partial_encoded_chunk` could be replaced by one which requires payload to be validated, but uses in `Client::process_partial_encoded_chunk` need to just get the payload out regardless of its state.

Reply:
I see